File: generator/src/main/java/ch/bailu/gtk/table/EnumTable.java
package ch.bailu.gtk.table;

import java.util.HashMap;
import java.util.Map;

import ch.bailu.gtk.converter.NamespaceType;

public class EnumTable {
    private final Map<String, Map<String, String>> table = new HashMap<>();


    private final static EnumTable INSTANCE = new EnumTable();

    public static EnumTable instance() {
        return INSTANCE;
    }

    private EnumTable() {}

    public void add(NamespaceType type) {
        if (type.isValid()) {
            getTable(type.getNamespace()).put(type.getName(), type.getName());
        }
    }

    private Map<String, String> getTable(String namespace) {
        Map<String, String> result = table.get(namespace);

        if (result == null) {
            result = new HashMap<>();
            table.put(namespace, result);
        }
        return result;
    }

    public boolean contains(NamespaceType type) {
        return type.isValid() && getTable(type.getNamespace()).containsKey(type.getName());
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/table/EnumTable.kt
package ch.bailu.gtk.table

import ch.bailu.gtk.model.type.NamespaceType
import ch.bailu.gtk.log.Logable
import ch.bailu.gtk.parser.tag.ParameterTag
import java.io.Writer
import java.util.*

object EnumTable: Logable {
    private val table: MutableMap<String, MutableMap<String, String>> = HashMap()

    fun add(type: NamespaceType) {
        if (type.valid) {
            getTable(type.namespace)[type.name] = type.name
        }
    }

    private fun getTable(namespace: String): MutableMap<String, String> {
        var result = table[namespace]
        if (result == null) {
            result = HashMap()
            table[namespace] = result
        }
        return result
    }

    operator fun contains(type: NamespaceType): Boolean {
        return type.valid && getTable(type.namespace).containsKey(type.name)
    }

    fun isEnum(namespace: String, parameter: ParameterTag): Boolean {
        return isEnum(namespace, parameter.getTypeName()) || isEnum(namespace, parameter.getType())
    }

    private fun isEnum(namespace: String, typeName: String): Boolean {
        return contains(NamespaceType(namespace, typeName))
    }

    override fun log(writer: Writer) {
        table.onEach { namespace ->
            writer.write("{${namespace.key}\n")
            namespace.value.forEach {
                writer.write(String.format("    %-40s\n", it.key))
            }
            writer.write("}\n\n")
        }
    }
}
